core/src/main/java/org/apache/druid/query/QueryException.java

```diff
@@ -30,12 +29,126 @@
 import java.util.function.Function;
 
 /**
- * Base serializable error response
- *
+ * Base serializable error response.
+ * <p>
+ * The Object Model that QueryException follows is a little non-intuitive as the primary way that a QueryException is
+ * generated is through a child class.  However, those child classes are *not* equivalent to a QueryException, instead
+ * they act as a Factory of QueryException objects.  This can be seen in two different places.
+ * <p>
+ * 1. When sanitize() is called, the response is a QueryException without any indication of which original exception
+ * occurred.
+ * 2. When these objects get serialized across the wire the recipient deserializes a QueryException. The client is
+ * never expected, and fundamentally is not allowed to, ever deserialize a child class of QueryException.
+ * <p>
+ * For this reason, QueryException must contain all potential state that any of its child classes could ever want to
+ * push across the wire.  Additionally, any catch clauses expecting one of the child Exceptions must know that it is
+ * running inside of code where the exception has not traveled across the wire.  If there is a chance that the
+ * exception could have been serialized across the wire, the code must catch a QueryException and check the errorCode
+ * instead.
+ * <p>
+ * As a correlary, adding new state or adjusting the logic of this class must always be done in a backwards-compatible
+ * fashion across all child classes of QueryException.
+ * <p>
+ * If there is any need to do different logic based on the type of error that has happened, the only reliable method
+ * of discerning the type of the error is to look at the errorCode String.  Because these Strings are considered part
+ * of the API, they are not allowed to change and must maintain their same semantics.  The known errorCode Strings
+ * are pulled together as public static fields on this class in order to make it more clear what options exist.
+ * <p>
  * QueryResource and SqlResource are expected to emit the JSON form of this object when errors happen.
  */
 public class QueryException extends RuntimeException implements SanitizableException
 {
+  /**
+   * Error codes
```

Comment:
Yeah...  Converting to and from the enum when serializing is a bit more complex as Jackson will want to serialize the enum values instead of the String, but our API is defined based on the Strings themselves.  In general, the `QueryException` concept itself probably needs a bit of love.  We could adjust it to have a numerical identifier for the type of the exception and then make the "code" just a description.  At that point we could codify the enum based on the numerical identifiers.  But that's an API migration as anybody relying on those codes would have to actually switch to rely on numbers instead.

Reply:
Yeah, numeric error codes would be nice to have.